### ISSUE TITLE
[java] UnnecessaryCast false positive with unchecked cast

### ISSUE THREAD

**Issue by maikelsteneker** [pmd/pmd#2748]
**Affects PMD Version:** 6.26.0

**Description:**

The rule [UnnecessaryCast](https://pmd.github.io/latest/pmd_rules_java_codestyle.html#unnecessarycast) tells me that I'm doing a cast that is not necessary, but omitting it results in a compiler error.

This refers to the cast of `(T[])list.toArray()`  from the example below. The `toArray()` methods returns an `Object[]`, not a `T[]`. There is an alternative method available that takes an array as argument and uses the type of that for its return value, but in a case like this, where `T` is a type parameter instead of a concrete type for which an array can be initialized, that alternative method can't be used either.

**Code Sample demonstrating the issue:**

```java
    /** Returns all elements from the given iterator in a new array */
    @SuppressWarnings("unchecked")
    public static <T> T[] asArray(Iterator<T> elements) {
        final List<T> list = new ArrayList<T>();
        while( elements.hasNext() ) list.add( elements.next() );
 
        T[] result = toArrayC( (Class<T>)null, list );
        if( result == null ) result = (T[])list.toArray();  // <----- this cast is really necessary
        return result;
    }

```

**Running PMD through:** *CLI*

**Comment by tweimer** [comment]
That cast `(T[])list.toArray()` will even cause a ClassCastException at runtime if `T` is not an `Object`. 

See the PMD rule `ClassCastExceptionWithToArray` for an example.

**Comment by oowekyala** [comment]
> That cast (T[])list.toArray() will even cause a ClassCastException at runtime if T is not an Object.

For reference, the cast is unchecked because T is erased to Object. So `(T[])` is erased to `(Object[])` and is a noop. But a caller might see a ClassCastException if they store the result in a reified array type that is not `Object[]`, eg `String[]`:
```java
Iterator<String> iter = Arrays.asList("a").iterator();
String[] array = asArray(iter); // boom
```

**Comment by adangel** [comment]
This has been fixed with PMD 7.0.0-rc1.
